art/attacks/poisoning/gradient_matching_attack.py

```diff
@@ -0,0 +1,575 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+# persons to whom the Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements Gradient Matching clean-label attacks (a.k.a. Witches' Brew) on Neural Networks.
+| Paper link: https://arxiv.org/abs/2009.02276
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Any, Dict, Tuple, TYPE_CHECKING, List
+
+import numpy as np
+from tqdm.auto import trange, tqdm
+
+from art.attacks.attack import Attack
+from art.estimators import BaseEstimator, NeuralNetworkMixin
+from art.estimators.classification.classifier import ClassifierMixin
+
+if TYPE_CHECKING:
+    # pylint: disable=C0412
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class GradientMatchingAttack(Attack):
```

Comment:
```suggestion
class GradientMatchingAttack(PoisoningAttackWhiteBox):
```

Reply:
I couldn't do this initially because of their different interface.
PoisoningAttackWhiteBox takes samples to poison and their target classes in poison method.
However, Gradient Matching attack needs (x_trigger, y_trigger) and (x_poison, y_poison). x_poison is samples to be manipulated, but the output changes for x_trigger instead, not x_poison. This can be confusing if we use the same interface, and Gradient Matching cannot replace other attacks implementing PoisoningAttackWhiteBox.